h/oauth/schemas.py

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+
+import colander
+
+from h import i18n
+from h.models.auth_client import GrantType, ResponseType
+from h.schemas.base import CSRFSchema
+
+_ = i18n.TranslationString
+
+
+def enum_schema_node(enum_cls, **kwargs):
+    """
+    Return a `colander.SchemaNode` for a field with a given enum type.
+
+    :param enum_cls: The enum class
+    :type enum_cls: enum.Enum
+    """
+    values = []
+    for name, _ in enum_cls.__members__.items():
+        values.append(name)
+
+    return colander.SchemaNode(colander.String(),
+                               validator=colander.OneOf(values),
+                               **kwargs)
+
+
+class CreateAuthClientSchema(CSRFSchema):
```

Comment:
There are some fairly complicated rules about which fields need to be set in certain circumstances. We should probably build this into the form at some point, not sure if now is the best time.

Some that we currently have:
* `redirect_uri` is required when `grant_type` is `authorization_code` (ensured with a database check constraint, so this renders a 500 right now.)
* `response_type` is required for certain `grant_type` values, this is validated in the authorize view code and might render a `unsupported_response_type` error.